diff --git a/tools/clang/scripts/build.py b/tools/clang/scripts/build.py
index 1abebf0092..edd2589612 100755
--- a/tools/clang/scripts/build.py
+++ b/tools/clang/scripts/build.py
@@ -191,7 +191,7 @@ def GetCommitDescription(commit):
   Needs to be called from inside the git repository dir."""
   git_exe = 'git.bat' if sys.platform.startswith('win') else 'git'
   return subprocess.check_output([
-      git_exe, 'describe', '--long', '--abbrev=8', '--match=*llvmorg-*-init',
+      'git', 'describe', '--long', '--abbrev=8', '--match=*llvmorg-*-init',
       commit
   ], universal_newlines=True).rstrip()