diff --git a/third_party/skia/gn/opts.gni b/third_party/skia/gn/opts.gni
index ce8724c892..9a727abe10 100644
--- a/third_party/skia/gn/opts.gni
+++ b/third_party/skia/gn/opts.gni
@@ -7,4 +7,5 @@
 _src = get_path_info("../src", "abspath")

 hsw = [ "$_src/opts/SkOpts_hsw.cpp" ]
-skx = [ "$_src/opts/SkOpts_skx.cpp" ]
+skx = [ ]
+# skx = [ "$_src/opts/SkOpts_skx.cpp" ]
diff --git a/third_party/skia/src/core/SkOpts.cpp b/third_party/skia/src/core/SkOpts.cpp
index 72a02f233f..f7ceff8bfa 100644
--- a/third_party/skia/src/core/SkOpts.cpp
+++ b/third_party/skia/src/core/SkOpts.cpp
@@ -45,7 +45,7 @@ namespace SkOpts {

     // Each Init_foo() is defined in src/opts/SkOpts_foo.cpp.
     void Init_hsw();
-    void Init_skx();
+    // void Init_skx();

     static bool init() {
     #if defined(SK_ENABLE_OPTIMIZE_SIZE)
@@ -55,9 +55,9 @@ namespace SkOpts {
             if (SkCpu::Supports(SkCpu::HSW)) { Init_hsw(); }
         #endif

-        #if (SK_CPU_SSE_LEVEL < SK_CPU_SSE_LEVEL_SKX) && defined(SK_ENABLE_AVX512_OPTS)
-            if (SkCpu::Supports(SkCpu::SKX)) { Init_skx(); }
-        #endif
+        // #if (SK_CPU_SSE_LEVEL < SK_CPU_SSE_LEVEL_SKX) && defined(SK_ENABLE_AVX512_OPTS)
+        //     if (SkCpu::Supports(SkCpu::SKX)) { Init_skx(); }
+        // #endif

     #endif
         return true;

