--- a/third_party/node/node.py
+++ b/third_party/node/node.py
@@ -24,7 +24,7 @@ def RunNode(cmd_parts, stdout=None):
   cmd = [GetBinaryPath()] + cmd_parts
   process = subprocess.Popen(
       cmd, cwd=os.getcwd(), stdout=subprocess.PIPE, stderr=subprocess.PIPE,
-      universal_newlines=True)
+      universal_newlines=True, encoding='utf-8')
   stdout, stderr = process.communicate()
 
   if process.returncode != 0:
